astarte/device/crypto.py

```diff
@@ -67,6 +68,17 @@ def import_device_certificate(client_crt, crypto_store_dir):
 
 
 def device_has_certificate(crypto_store_dir):
-    return path.exists(path.join(crypto_store_dir,
-                                 "device.crt")) and path.exists(
-                                     path.join(crypto_store_dir, "device.key"))
+    cert_path = path.join(crypto_store_dir, "device.crt")
+    key_path = path.join(crypto_store_dir, "device.key")
+
+    return path.exists(cert_path) and path.exists(key_path) and not certificate_is_expired(crypto_store_dir)
+
+
+def certificate_is_expired(crypto_store_dir):
+    cert_path = path.join(crypto_store_dir, "device.crt")
+    with open(cert_path, 'r') as file:
+        data = file.read()
+    if data:
+        certificate = x509.load_pem_x509_certificate(data.encode('ascii'), default_backend())
+        return certificate.not_valid_after < datetime.utcnow()
+    return True
```

Comment:
This returns `True` when `data` is falsy, I assume it should return `False` in that case instead

Reply:
Also, how does `load_pem_x509_certificate` fail if `data` is not a valid x509 certificate? You should catch the exception and return False also in that case so the certificate gets recreated

Reply:
> This returns `True` when `data` is falsy, I assume it should return `False` in that case instead

The signature of `certificate_is_expired` says it checks if the certificate is expired, so a `True` is probably the better return, as False can be interpreted as a valid certificate.
Anyway I agree that the check on `data` is redundant as it is already checked by `device_has_certificate` before calling `certificate_is_expired`. I can remove the check in `certificate_is_expired` if it could help readability.

Reply:
> Also, how does `load_pem_x509_certificate` fail if `data` is not a valid x509 certificate? You should catch the exception and return False also in that case so the certificate gets recreated

Yes, you are right, according to the `load_pem_x509_certificate` function it would rise a ValueError we could catch.